### PR TITLE
ENH: optimize: add stagnation check to df-sane

### PR DESCRIPTION
**Pull request by pv**
Add a simple stagnation check to DF-SANE for bailing out early.

### CODE REVIEW

File: scipy/optimize/_spectral.py [hunk 1]
```diff
@@ -58,7 +59,11 @@ def _root_df_sane(func, x0, args=(), ftol=1e-8, fatol=1e-300, maxfev=1000,
     M : int, optional
         Number of iterates to include in the 'cruz' nonmonotonic line search.
         Default: 10
-    line_search : {'cruz', 'cheng'}
+    stagnation_limit : int, optional
+        If merit function does not decrease in `stagnation_limit` iterations,
+        terminate with failure.
+        Default: 100
```
Comment (larsmans) [inline scipy/optimize/_spectral.py]: I've seen this option being called "patience" in other contexts (backpropagation in particular). Wouldn't that be a better name, it being easier to type?

Comment (josef-pkt) [inline scipy/optimize/_spectral.py]: for me `stagnation_limit` is more informative and would be easier to remember than `patience` which I've never heard before. patience for what?

Comment (larsmans) [inline scipy/optimize/_spectral.py]: Patience before terminating for lack of progress, I guess.

File: scipy/optimize/_spectral.py [hunk 2]
```diff
@@ -122,12 +131,21 @@ def fmerit(F):
...
             callback(x_k, F_k)
 
+        # Check convergence
         if F_k_norm < ftol * F_0_norm + fatol:
             # Converged!
             message = "successful convergence"
             converged = True
             break
 
+        if f_k < last_improved_f:
```
Comment (pv) [inline scipy/optimize/_spectral.py]: This probably should be something like `f_k < 0.99*last_improved_f` to enforce at least linear convergence...

Comment (argriffing) [inline scipy/optimize/_spectral.py]: Is this still active?  I don't have an opinion on the `stagnation_limit` vs. `patience` naming issue.

Comment (dlax) [inline scipy/optimize/_spectral.py]: Added a needs-work label until this gets addressed.

**Comment by ev-br** [comment]
It's a bit sad to see this linger because of the name of a parameter

**Comment by dlax** [comment]
Evgeni Burovski wrote:
> It's a bit sad to see this linger because of the name of a parameter


I added the "needs-work" label because of 
https://github.com/scipy/scipy/pull/4459#discussion_r23557464, not 
because of the naming "issue" (I find `stagnation_limit` fine, BTW).

**Comment by pv** [comment]
There's also the `0.99*last_improved_f` thing. IIRC I have better version somewhere but didn't get back around to doing it yet...

**Comment by mdhaber** [comment]
@pv I'd like to merge this, but I need to merge main. Can you "[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)"?

**Comment by mdhaber** [comment]
@pv would you be willing to "[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests)" here?

**Comment by mdhaber** [comment]
In that case, I'll go ahead and close this. The code was mostly complete, so if someone wants to review a new PR with the same content, let me know and I'll submit one.

**Pull request closed without merging**